Introducing vault sharing and the restrictions of pulling and cloning with vault permissions

CMCDragonkai opened the pull request:
### Description

This pull request aims to allow vaults to be shared and unshared with other Node Ids. Specification and discussion can be found inside the Tasklist issues.

### Secret Commands

There are some changes to secrets commands here. These should be extracted out of this PR and moved to another PR as they have different specification requirements.

### Issues Fixed

* Fixes MatrixAI/Polykey#259 
* Fixes MatrixAI/Polykey#253 
* Fixes MatrixAI/Polykey#305
* Fixes MatrixAI/Polykey#299 
* Fixes MatrixAI/Polykey#309
* Fixes MatrixAI/Polykey#335 
* Fixes MatrixAI/Polykey#336
* Fixes MatrixAI/Polykey#337 
* Fixes MatrixAI/Polykey#338
* Fixes MatrixAI/Polykey#339
* Fixes MatrixAI/Polykey#340
* Fixes MatrixAI/Polykey#342 
* Related to MatrixAI/Polykey#294 - in terms of introducing withF and withG to be used
* Related to MatrixAI/Polykey-Docs#3 - review sharing diagrams when changes are done

### Tasks

Reference for residual TODOs is here https://github.com/MatrixAI/js-polykey/pull/266#issuecomment-994242267 .

- [x] 6. MatrixAI/Polykey#337 
- [x] 7. MatrixAI/Polykey#338
- [x] 9. MatrixAI/Polykey#340
- [x] 2. MatrixAI/Polykey#342 
- [x] 8. MatrixAI/Polykey#339
- [x] 1. MatrixAI/Polykey#305  
- [x] 3. MatrixAI/Polykey#259 
- [x] 4. MatrixAI/Polykey#335 
- [x] 5. MatrixAI/Polykey#336

### Final checklist
* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

scottmmorris commented on src/agent/utils.ts:
```diff
@@ -29,6 +29,7 @@ async function checkAgentRunning(nodePath: string): Promise<boolean> {
 async function spawnBackgroundAgent( // FIXME, this is broken.
```
In this branch the tests for spawning an agent are passing so I don't know if this is still relevant

scottmmorris replied:
I've removed this

scottmmorris commented on src/bin/agent/start.ts:
```diff
@@ -14,6 +14,7 @@ const start = binUtils.createCommand('start', {
   format: true,
   passwordFile: true,
 });
+start.option('-fr, --fresh', 'Starts a fresh agent');
```
There is probably a better way to specify the fresh option but considering the CLI commands are changing soon I've left that up to future changes

scottmmorris commented on src/vaults/VaultInternal.ts:
```diff
@@ -48,44 +48,45 @@ class VaultInternal {
         },
         message: 'Initial Commit',
       });
       await efs.writeFile(
```
In another PR I might have a look at the need to write this file here. I remember that cloning and pulling with iso-git fails without (and for some reason it isnt automatically written). This may be because we are looking for it in the git utility functions rather than it being a requirement in the iso-git library. See the `packedrefs` function in `git/utils.ts`

scottmmorris replied:
https://github.com/MatrixAI/js-polykey/issues/298

joshuakarp commented on src/vaults/VaultInternal.ts:
```diff
@@ -25,94 +25,96 @@ class VaultInternal {
   public readonly baseDir: string;
   public readonly gitDir: string;
   public readonly vaultId: VaultId;
+  protected readonly remote: boolean;
```
From discussions here https://github.com/MatrixAI/js-polykey/issues/253#issuecomment-966704561, I was expecting this to be some kind of `Remote` type. Is there a reason why this is a boolean instead now?

scottmmorris replied:
I realised it wouldn't really make a huge difference. I looked at it but it would require some reshuffling to the db. On a similar vein their could be some more restructuring  of the vault metadata that is stored in the db and doing this could be a small optimisation for the future. Ill see if I have time for this tomorrow to implement but the gains from it wouldn't be hugely significant

scottmmorris replied:
Here Ive just made it simple by settting the remote field when constructed which is simply done through the Vault Manager. This is probably the most robust way for now but when expanding the COW and further systems for the vaults this would have to be changed

tegefaulkes replied:
@joshuakarp can this be resolved?

joshuakarp replied:
Yeah, this can be resolved now. I've added an issue to potentially consider usage of this at a later date: #309

joshuakarp commented on tests/bin/echo.test.ts:
```diff
@@ -0,0 +1,80 @@
+import * as testUtils from './utils';
```
This entire file can likely be removed now, as I believe `echo` has been removed.

joshuakarp commented on tests/client/clientService.test.ts:
```diff
@@ -0,0 +1,2274 @@
+import type { Host, Port, TLSConfig } from '@/network/types';
```
This file can also be removed, seeing as @scottmmorris has separated this in master now.

joshuakarp commented on src/keys/KeyManager.ts:
```diff
@@ -364,11 +424,11 @@ class KeyManager {
     subjectAttrsExtra: Array<{ name: string; value: string }> = [],
     issuerAttrsExtra: Array<{ name: string; value: string }> = [],
-  ): Promise<void> {
+  ): Promise<RecoveryCode> {
```
As an interesting point here, I noticed that while rebasing, it seemed to favour the original `): Promise<void> {`  from `master` (as opposed to the change made here) without showing any kind of conflict.

CMCDragonkai replied:
It should be `Promise<void>`. Recovery code is acquired by `getRecoveryCode()`.

joshuakarp replied:
Reverting this return type to `void` (as well as `resetRootKeyPair`) based on discussion with @CMCDragonkai. We have a `getRecoveryCode` function that performs this for us. There's no instance where we require the function to explicitly return recovery code.

CMCDragonkai commented on src/agent/GRPCClientAgent.ts:
```diff
@@ -88,14 +88,17 @@ class GRPCClientAgent extends GRPCClient<AgentServiceClient> {
   @ready(new agentErrors.ErrorAgentClientDestroyed())
-  public vaultsScan(...args) {
-    return grpcUtils.promisifyReadableStreamCall<vaultsPB.Vault>(
+  public nodesScan(...args) {
```
Should be `vaultsScan` with the ability to take gestalt id.

CMCDragonkai replied:
Also moved to the right location.

CMCDragonkai commented on src/agent/agentService.ts:
```diff
@@ -98,80 +133,88 @@ function createAgentService({
     },
-    vaultsScan: async (
-      call: grpc.ServerWritableStream<nodesPB.Node, vaultsPB.Vault>,
+    nodesScan: async (
```
Should be `vaultsScan` with the ability to take gestalt id.

tegefaulkes replied:
As far as I know this just scans a single node and not the whole gestalt. gestaltId applicable here?

CMCDragonkai replied:
It can be a gestalt ID in the future. For now stick with a node Id. Because you can scan the entire target gestalt by broadcasting it and following the crytolinks in the gestalt graph.

CMCDragonkai commented on src/bin/nodes/CommandFind.ts:
```diff
@@ -87,7 +87,7 @@ class CommandFind extends CommandPolykey {
         // Like ping it should error when failing to find node for automation reasons.
         if (!result.success)
-          throw new CLIErrors.ErrorNodeFindFailed(result.message);
+          throw new CLIErrors.ErrorCLINodeFindFailed(result.message);
```
If you cannot resolve a node id on the DHT, doesn't it just loop forever? Therefore by default, the command can just run forever looking for the node until it appears. No need to throw an error here.

Furthermore if there is a timeout applied, which should be an option, then this is domain error of `nodes`, not `CLI`.

tegefaulkes replied:
I think having it go forever and having it timeout both have utility. I'll add an option for a timeout.

CMCDragonkai replied:
Yea it's just default behaviour it can just run forever. Then optionally the user can add a timeout. These options should be put into `bin/utils/options.ts`.

tegefaulkes replied:
It now defaults to trying forever. I added two options to specify a timeout, you can specify how many times to attempt and the delay between attempts.

```ts
const attemptsCount = new commander.Option(
  '-c --count <count>',
  'Number of attempts before failing'
)
  .argParser(binParsers.parseNumber);

const attemptsDelay = new commander.Option(
  '-ad --attempt-delay <delay>',
  'Number of milliseconds between each attempt'
)
  .argParser(binParsers.parseNumber)
  .default(5000);
  ```

CMCDragonkai replied:
Cool but try to use the same names we've used elsewhere. `--retry X` and `--retry-interval` or `--retry-timeout`.

CMCDragonkai commented on src/bin/nodes/CommandNodes.ts:
```diff
@@ -13,6 +14,7 @@ class CommandNodes extends CommandPolykey {
     this.addCommand(new CommandClaim(...args));
     this.addCommand(new CommandFind(...args));
     this.addCommand(new CommandPing(...args));
+    this.addCommand(new CommandScan(...args));
```
Move back to `vaults`.

CMCDragonkai commented on src/bin/nodes/CommandPing.ts:
```diff
@@ -53,7 +53,7 @@ class CommandPing extends CommandPolykey {
         } catch (err) {
           if (err instanceof nodesErrors.ErrorNodeGraphNodeNotFound) {
-            error = new CLIErrors.ErrorNodePingFailed(
+            error = new CLIErrors.ErrorCLINodePingFailed(
```
Same situation with `find`. If `ping` doesn't respond, you just wait forever.

It seems like ping would basically do find first and then try to ping the node. So it may wait forever to find/resolve, and then wait forever to ping.

An optional parameter for timing out should be available.

tegefaulkes replied:
Added the same options as CommandFind

CMCDragonkai commented on src/bin/nodes/CommandScan.ts:
```diff
@@ -0,0 +1,71 @@
+import type { Metadata } from '@grpc/grpc-js';
```
Move to `bin/vaults`.

CMCDragonkai commented on src/bin/vaults/CommandPermissions.ts:
```diff
@@ -1,115 +0,0 @@
-// Import Logger, { LogLevel, StreamHandler } from '@matrixai/logger';
```
Uncomment this and keep it in `CommandVaults`. Should be similar to `identities/CommandPermissions`.

Note that we can later make it work in `pk acl permissions <vaultid/identityid/nodeid>` something like that.

tegefaulkes replied:
Why are we doing this? all functionality for this has been stripped out or moved elsewhere.

CMCDragonkai replied:
It's only to reimplement `pk vaults permissions` command.

tegefaulkes replied:
I have re implemented the vaults permissions command.

CMCDragonkai commented on src/bin/vaults/CommandStat.ts:
```diff
@@ -1,98 +0,0 @@
-//     const grpcClient = client.grpcClient;
-//     const responseMessage = await binUtils.unaryCallCARL<vaultsPB.Stat>(
-//       client,
-//       attemptUnaryCall(client, grpcClient.vaultsSecretsStat),
```
Move this into `bin/secrets`. This should be `pk secrets stat` call that works like an fs stat call.

tegefaulkes replied:
Added.

CMCDragonkai commented on src/bin/errors.ts:
```diff
@@ -88,9 +72,6 @@ export {
-  ErrorNodeFindFailed,
-  ErrorNodePingFailed,
+  ErrorCLINodeFindFailed,
+  ErrorCLINodePingFailed,
```
Shouldn't be necessary.

tegefaulkes replied:
Keeping these as the commands still fail with these errors when a timeout is set.

CMCDragonkai replied:
But but why are these not domain errors?

CMCDragonkai commented on src/PolykeyAgent.ts:
```diff
@@ -260,30 +260,32 @@ class PolykeyAgent {
         nodeManager,
         logger: logger.getChild(Discovery.name),
       }));
+
```
Remove extra line.

CMCDragonkai commented on src/bin/vaults/CommandPull.ts:
```diff
@@ -9,12 +9,19 @@ class CommandPull extends CommandPolykey {
+    this.option(
+      '-pv, --pull-vault [pullVaultNameOrId]',
+      'Name or Id of the vault to pull from',
+    );
```
Move these into `bin/utils/options.ts`. They should be standardised in our bin.

tegefaulkes replied:
Fixed.

CMCDragonkai commented on src/bin/vaults/CommandScan.ts:
```diff
@@ -1,109 +0,0 @@
-// Import Logger, { LogLevel, StreamHandler } from '@matrixai/logger';
```
Uncomment and enable this command, and test it.

CMCDragonkai commented on src/client/GRPCClientClient.ts:
```diff
@@ -167,34 +167,26 @@ class GRPCClientClient extends GRPCClient<ClientServiceClient> {
 
   @ready(new clientErrors.ErrorClientClientDestroyed())
-  public vaultsScan(...args) {
+  public nodesScan(...args) {
```
Should be `vaultsScan`.

CMCDragonkai commented on src/client/rpcNodes.ts:
```diff
@@ -149,6 +152,26 @@ const createNodesRPC = ({
         return;
       }
     },
+    nodesScan: async (
```
Should be `vaultsScan` and obviously in `rpcVaults`.

CMCDragonkai commented on src/git/utils.ts:
```diff
@@ -152,14 +161,30 @@ function textToPackedRefs(text: string): Refs {
+ * @param fs
+ * @param gitdir
+ * @param filepath
+ */
```
These sorts of comments are not necessary. TS function signature is already sufficient. Unless you are adding additional commentary to the options.

CMCDragonkai commented on src/keys/KeyManager.ts:
```diff
@@ -38,28 +37,26 @@ class KeyManager {
   public readonly rootCertsPath: string;
   public readonly dbKeyPath: string;
-  public readonly vaultKeyPath: string;
+  public readonly litmusMessagePath: string;
```
I'm pretty sure this was meant to be called the `canary`. It can just be `canaryPath`.

It was meant to tell us if our descryption is successful by first encrypting a known static value, and then descrypting to see if it is the known static value.

It should have it's own custom exception if it failed in the `keys/errors.ts` and then trigger failure on `pk bootstrap` and `pk agent start`.

tegefaulkes replied:
From what I can tell this shouldn't exist anymore. might of snuck in when the vault key was removed. I'll remove it.

CMCDragonkai commented on src/keys/KeyManager.ts:
```diff
@@ -174,8 +166,7 @@ class KeyManager {
     this._dbKey = await this.setupKey(this.dbKeyPath, this.dbKeyBits);
-    this._vaultKey = await this.setupKey(this.vaultKeyPath, this.vaultKeyBits);
-    this.logger.info(`Started ${this.constructor.name}`);
+    this.logger.info('Created Key Manager');
```
This is incorrect, replace it with `this.logger.info(`Started ${this.constructor.name}`);`.

CMCDragonkai commented on src/keys/KeyManager.ts:
```diff
@@ -38,28 +37,26 @@ class KeyManager {
   protected rootCertDuration: number;
   protected dbKeyBits: number;
-  protected vaultKeyBits: number;
+  protected _generatedRecoveryCode: RecoveryCode;
```
What's this for? I thought we resolved the recovery code issues in `master`.

CMCDragonkai commented on src/keys/KeyManager.ts:
```diff
@@ -120,11 +113,10 @@ class KeyManager {
     this.rootCertsPath = path.join(keysPath, 'root_certs');
     this.dbKeyPath = path.join(keysPath, 'db.key');
-    this.vaultKeyPath = path.join(keysPath, 'vault.key');
+    this.litmusMessagePath = path.join(keysPath, 'litmusMessage');
```
Is this being used atm? I think it was forgotten to be implemented...? If it was done elsewhere, it should be done here, not in vaults.

CMCDragonkai replied:
I remember talking about this with @scottmmorris before. It was intended to be `canaryPath`. It just seems it was never fully implemented. If so, then you can just remove it, and we can add it in later feature.

CMCDragonkai commented on src/proto/schemas/polykey/v1/client_service.proto:
```diff
@@ -28,6 +28,7 @@ service ClientService {
   rpc NodesPing(polykey.v1.nodes.Node) returns (polykey.v1.utils.StatusMessage);
   rpc NodesClaim(polykey.v1.nodes.Claim) returns (polykey.v1.utils.StatusMessage);
   rpc NodesFind(polykey.v1.nodes.Node) returns (polykey.v1.nodes.NodeAddress);
+  rpc NodesScan(polykey.v1.nodes.Node) returns (stream polykey.v1.vaults.List);
```
Should be `VaultsScan`.

CMCDragonkai commented on src/nodes/NodeManager.ts:
```diff
@@ -600,11 +601,13 @@ class NodeManager {
    */
   @ready(new nodesErrors.ErrorNodeManagerNotRunning())
-  public async scanNodeVaults(nodeId: string): Promise<Array<string>> {
+  public async scanNodeVaults(
```
Not sure if we are still proxying all calls. But I guess this will be fixed in our node management refactoring #310

CMCDragonkai commented on src/sigchain/Sigchain.ts:
```diff
@@ -220,7 +220,9 @@ class Sigchain {
    */
   @ready(new sigchainErrors.ErrorSigchainNotRunning())
-  public async addClaim(claimData: ClaimData): Promise<ClaimEncoded> {
+  public async addClaim(
```
Were these brought in from @emmacasolin in #311

emmacasolin replied:
I made `addClaim()` return a `ClaimEncoded` in https://github.com/MatrixAI/js-polykey/pull/278 in order to implement the `identities claim` command, and then `ClaimId` was added here https://github.com/MatrixAI/js-polykey/pull/292/commits/3dea67340cb83a4e385a934a6f4323a67f19096c. But we were talking on Slack about how probably only `ClaimId` needs to be returned.

tegefaulkes replied:
I pulled it in from qa-testing to fix a problem when I matched dependency versions with qa-testing.

CMCDragonkai reviewed (commented):
These things still need to be fixed up before merging.

CMCDragonkai commented on src/client/service/vaultsScan.ts:
```diff
@@ -1,40 +1,37 @@
-      const vaults = await vaultManager.listVaults();
-      vaults.forEach(async (vaultId, vaultName) => {
+      const nodeId = nodesUtils.makeNodeId(call.request.getNodeId());
+      const list = await nodeManager.scanNodeVaults(nodeId);
```
@tegefaulkes this should still be part of the `vaultManager` and not `nodeManager`.

tegefaulkes replied:
This is being addressed in #310.

CMCDragonkai replied:
So if #310 moves this feature into `vaultManager`, I think that `VaultManager` will be using the `NodeConnection` features to actually call the scan vaults. Basically there shouldn't be a method called `scanNodeVaults` at all in `NodeManager`. If you're removing that code though, make sure to post it as as comment here, so I can refer to it.

CMCDragonkai committed:
Vault sharing, permissions and scanning:
- NodeId fixes for vaults tests
- Replaced makeVaultIdPretty` with `encodeVaultId`

tegefaulkes committed:
Fixes #305

tegefaulkes committed:
Fixes #337
- `VaultId` usage,
- updated validation
- removal of `GenericIdTypes.ts`

tegefaulkes committed:
Fixes #340
- selective imports
- error descriptions updated
- Using google `Timestamp` message in `LogEntry` message
- fixing comments

tegefaulkes committed:
Fixes #339 - Vaults scanning

tegefaulkes committed:
Fixes #342 - added `connectionInfoGetter` utility

tegefaulkes committed:
Fixes #259 replacing `un/shareVault` with `vaultsPermissionSet`, `vaultsPermissionGet` and `vaultsPermissionUnset`.

tegefaulkes committed:
Fixes #335
- prototyping committing history
- prototyping committing, EOD
- Linear and branching committing implemented
- recovery from dirty state and expanding tests
- added global garbage collection for git objects.

tegefaulkes committed:
Fixes #336

tegefaulkes committed:
General fixes
